File: app/src/main/java/org/evcs/android/features/profile/wallet/PaymentMethodAdapterV2.java
package org.evcs.android.features.profile.wallet;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;

import org.evcs.android.R;
import org.evcs.android.model.PaymentMethod;
import org.evcs.android.ui.adapter.BaseRecyclerAdapter;
import org.evcs.android.util.UserUtils;

public class PaymentMethodAdapterV2 extends BaseRecyclerAdapter<PaymentMethod, PaymentMethodViewHolderV2> {

    private CreditCardListener mListener;
    private String mDefaultPm;

    public PaymentMethodAdapterV2() {
        mDefaultPm = UserUtils.getLoggedUser().defaultPm;
    }

    @Override
    protected void populate(PaymentMethodViewHolderV2 holder, final PaymentMethod item,
                            int position) {
        holder.setPaymentMethod(item.card);
        holder.setDefault(item.id.equals(mDefaultPm));
        holder.setListeners(new CreditCardView.CreditCardViewListener() {
            @Override
            public void onStarClicked() {
                if (mListener != null) mListener.onStarClicked(item);
            }

            @Override
            public void onTrashClicked() {
                if (mListener != null) mListener.onTrashClicked(item);
            }
        });
    }

    public void setOnItemClickListener(CreditCardListener listener) {
        mListener = listener;
    }

    @Override
    @NonNull
    public PaymentMethodViewHolderV2 onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater
            .from(parent.getContext()).inflate(R.layout.view_credit_card_wrapper, parent, false);
        return new PaymentMethodViewHolderV2(v);
    }

    public void setDefault(PaymentMethod item) {
        mDefaultPm = item.id;
        notifyDataSetChanged();
    }

//    public CreditCard getDefault() {
//        for (int i = 0; i < getItemCount(); i++) {
//            if (get(i).card.isDefault()) return get(i).card;
//        }
//        return null;
//    }

    public interface CreditCardListener {
        void onStarClicked(PaymentMethod item);
        void onTrashClicked(PaymentMethod item);
    }
}

File: app/src/main/java/org/evcs/android/features/profile/wallet/PaymentMethodAdapterV2.kt
package org.evcs.android.features.profile.wallet

import org.evcs.android.ui.adapter.BaseRecyclerAdapter
import org.evcs.android.util.UserUtils
import android.view.ViewGroup
import android.view.LayoutInflater
import org.evcs.android.R
import org.evcs.android.model.PaymentMethod

class PaymentMethodAdapterV2 : BaseRecyclerAdapter<PaymentMethod, PaymentMethodViewHolder>() {

    private var mListener: CreditCardListener? = null
    private var mDefaultPm: String? = null

    init {
        mDefaultPm = UserUtils.getLoggedUser().defaultPm
    }

    override fun populate(holder: PaymentMethodViewHolder, item: PaymentMethod, position: Int) {
        holder.setPaymentMethod(item)
        holder.setDefault(item.id == mDefaultPm)
        holder.setListeners(object : PaymentMethodViewHolder.CreditCardViewListener {
            override fun onStarClicked() {
                if (mListener != null) mListener!!.onStarClicked(item)
            }

//            override fun onTrashClicked() {
//                if (mListener != null) mListener!!.onTrashClicked(item)
//            }

            override fun onDetailClicked() {
                if (mListener != null) mListener!!.onDetailClicked(item)
            }
        })
    }

    fun setOnItemClickListener(listener: CreditCardListener?) {
        mListener = listener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PaymentMethodViewHolder {
        val v = LayoutInflater
            .from(parent.context).inflate(R.layout.view_credit_card_item, parent, false)
        return PaymentMethodViewHolder(v)
    }

    fun setDefault(id: String) {
        mDefaultPm = id
        notifyDataSetChanged()
    }

    fun setDefault(item: PaymentMethod) {
        setDefault(item.id!!)
    }

    fun getDefault(): PaymentMethod? {
        //Ugly
        val items = IntRange(0, itemCount - 1).map { i -> get(i) }
        return items.firstOrNull { item -> item.id == mDefaultPm }
    }

    interface CreditCardListener {
        fun onStarClicked(item: PaymentMethod)
//        fun onTrashClicked(item: PaymentMethod)
        fun onDetailClicked(item: PaymentMethod)
    }
}